File: hackerrank/algorithms/sorting/insertion-sort-advanced-analysis/insertion-sort-advanced-analysis.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

# HackerRank: Insertion Sort Advanced Analysis
# https://www.hackerrank.com/challenges/insertion-sort/
# By instrumenting recursive top-down mergesort.

# Instrumented mutating mergesort extensions for arrays.
class Array
  # Mergesorts this array. Returns the number of swaps insertion sort would do.
  def mergesort
    do_mergesort(0, size, [])
  end

  private

  def do_mergesort(low, high, aux)
    return 0 if high - low < 2

    mid = (low + high) / 2
    acc = do_mergesort(low, mid, aux) + do_mergesort(mid, high, aux)
    acc + merge(low, mid, high, aux)
  end

  def merge(low, mid, high, aux)
    raise 'auxiliary array expected to be empty' unless aux.empty?

    acc = 0
    left = low
    right = mid

    while left < mid && right < high
      if self[right] < self[left]
        aux << self[right]
        right += 1
        acc += mid - left # Number of elems insertion sort would swap across.
      else
        aux << self[left]
        left += 1
      end
    end

    while left < mid
      aux << self[left]
      left += 1
    end

    raise 'bug: auxiliary storage wrongly populated' if aux.size != right - low
    self[low...right] = aux
    aux.clear

    acc
  end
end

def read_value
  gets.to_i
end

def read_record
  length = read_value
  record = gets.split.map(&:to_i)
  raise 'wrong record length' if record.size != length
  record
end

read_value.times { puts read_record.mergesort } if $PROGRAM_NAME == __FILE__

File: hackerrank/algorithms/sorting/insertion-sort-advanced-analysis/insertion-sort-advanced-analysis.java
// HackerRank: Insertion Sort Advanced Analysis
// https://www.hackerrank.com/challenges/insertion-sort/
// By instrumenting recursive top-down mergesort.

import java.util.Scanner;

enum Algo {
    ;

    /**
     * Mergesorts an array and returns the number of inversions it had.
     * @param{values}  The array to sort and count inversions in.
     * @return  The number of inversions originally present. This is the same
     *          as the number of swaps/shifts insertion sort would do.
     * */
    static int mergesort(int[] values) {
        var aux = new int[values.length];
        return mergesortSubarray(values, 0, values.length, aux);
    }

    private static int
    mergesortSubarray(int[] values, int low, int high, int[] aux) {
        if (high - low < 2) return 0;

        var mid = low + (high - low) / 2;

        var count = 0;
        count += mergesortSubarray(values, low, mid, aux);
        count += mergesortSubarray(values, mid, high, aux);
        count += merge(values, low, mid, high, aux);
        return count;
    }

    private static int
    merge(int[] values, int low, int mid, int high, int[] aux) {
        var count = 0;
        var left = low;
        var right = mid;
        var out = 0;

        while (left < mid && right < high) {
            if (values[right] < values[left]) {
                aux[out++] = values[right++];
                count += mid - left;
            } else {
                aux[out++] = values[left++];
            }
        }

        while (left < mid) aux[out++] = values[left++];
        if (right - low != out) throw new AssertionError("length mismatch");
        for (var i = 0; i < out; ++i) values[low++] = aux[i];
        return count;
    }
}

enum Solution {
    ;

    public static void main(String[] args) {
        try (Scanner sc = new Scanner(System.in)) {
            for (var t = sc.nextInt(); t > -0; --t) {
                System.out.println(Algo.mergesort(readRecord(sc)));
            }
        }
    }

    private static int[] readRecord(Scanner sc) {
        var values = new int[sc.nextInt()];
        for (var i = 0; i < values.length; ++i) values[i] = sc.nextInt();
        return values;
    }
}
